### DISCUSSION TITLE
Questions about the pool and resetting the session state.

### DISCUSSION THREAD

**Discussion by paulzakin** [psycopg/psycopg#893] [Q&A] [answered]
Hello!

Two quick questions:

1. I  looked through the documentation and could not find any reference about the pool resetting any state created for the session. In PgBouncer and PostgreSQL it would be something like DISCARD ALL.

2. If two, different Python applications on the same server (in the same Docker container, for what that is worth) each uses a pool, I assume there is no chance of “cross-pollination” in the sense that application 2 can use the connections held by application 1. This is mostly a sanity check for me.

Thanks!

**Comment by dvarrazzo** [comment]
Hello,

no, the connection pool, by default, doesn't reset the connection, apart from discarding an eventual open transaction or broken connection. If you need a `DISCARD ALL`, you can execute it in a `reset` callback.

The pool is not designed for "adversarial" usage, it assumes that you will use it within your application and you configure it appropriately. If you want to use it from two different apps you should configure an appropriate `reset`, but then you cannot use a specific `setup`.

All in all it would be safer and more efficient to use two pools, IMO.

**Reply by paulzakin** [reply]
1. Makes sense on the DISCARD ALL.
2. Ok, just so I understand, if each unique Python process (e.g. Celery and Django) on the same server instantiates their own pool to connect to the same database - each pool should have their own set of connections and not share it with the other one? The “not sharing” is the desired behavior for sure and what I assumed is the behavior. But just wanted to check!

**Reply by dvarrazzo** [reply]
The pool is an in-process pool: you can instantiate one in a process and several threads or asyncio tasks can use it concurrently, but it cannot be shared across processes.

So it's not like pgbouncer, which is a service and other external services would get connections from it.

Be careful about celery because it forks internally. You cannot have a pool in the main process and use it from forked processes (because libpq connections cannot be passed across processes). For django, it depends on the app server you use: if you use a multithread one instead of a forking one it will work.

**Reply by paulzakin** [reply]
Ah, the Celery was just an example - we don’t actually use it. So - what I think you are saying is the pool (and the connections inside of it?) are isolated across processes. However, the pool (and the connections inside of it) may not be isolated within a process depending on whether threads or forking is used? 

In short, there is no way for Django to consume connections from the pool that Celery is using and vice versa, as they are separate processes. But within Django or Celery, it would depend. That is now my understanding at least! 

Really appreciate you answering these questions!

**Reply by dvarrazzo** [reply, marked as answer]
> Ah, the Celery was just an example - we don’t actually use it. So - what I think you are saying is the pool (and the connections inside of it?) are isolated across processes. However, the pool (and the connections inside of it) may not be isolated within a process depending on whether threads or forking is used?

Yes, this is correct.

> In short, there is no way for Django to consume connections from the pool that Celery is using and vice versa, as they are separate processes. But within Django or Celery, it would depend. That is now my understanding at least!

If Django and Celery in the example are two separate processes, they won't share connections. The connections are acquired by the respective pools in the two different processes and will be exclusive to them.

**Reply by paulzakin** [reply]
Thanks so much!!
